contracts/margin/external/ShortEthOpener.sol

```diff
@@ -0,0 +1,154 @@
+
+    // ============ State Variables ============
+
+    address public DYDX_MARGIN;
+
+    address public WETH;
+
+    // ============ Constructor ============
+
+    constructor(
+        address margin,
+        address weth
+    )
+        public
+    {
+        DYDX_MARGIN = margin;
+        WETH = weth;
+
+        // WETH approval of maxInt does not decrease
+        WETH.approve(DYDX_MARGIN, MathHelpers.maxUint256());
+    }
+
+    // ============ Public Functions ============
+
+    /**
+     * Fallback function. Disallows ether to be sent to this contract without data except when
+     * unwrapping WETH.
+     */
+    function ()
+        external
+        payable
+    {
+        require(
+            msg.sender == WETH,
+            "ShortEthOpener#fallback: Cannot recieve ETH directly except when unwrapping WETH"
+        );
+    }
+
+    /**
+     * Increase the size of a position. Funds will be borrowed from the loan payer and sold as per
+     * the position. The amount of owedToken borrowed from the lender will be >= the amount of
+     * principal added, as it will incorporate interest already earned by the position so far.
+     *
+     * @param  addresses           Addresses corresponding to:
+     *
+     *  [0]  = loan payer
+     *  [1]  = loan taker
+     *  [2]  = loan position owner
+     *  [3]  = loan fee recipient
+     *  [4]  = loan lender fee token
+     *  [5]  = loan taker fee token
+     *  [6]  = exchange wrapper address
+     *
+     * @param  values256           Values corresponding to:
+     *
+     *  [0]  = loan maximum amount
+     *  [1]  = loan minimum amount
+     *  [2]  = loan minimum heldToken
+     *  [3]  = loan lender fee
+     *  [4]  = loan taker fee
+     *  [5]  = loan expiration timestamp (in seconds)
+     *  [6]  = loan salt
+     *  [7]  = amount of principal to add to the position (NOTE: the amount pulled from the lender
+     *                                                           will be >= this amount)
+     *
+     * @param  values32            Values corresponding to:
+     *
+     *  [0] = loan call time limit (in seconds)
+     *  [1] = loan maxDuration (in seconds)
+     *
+     * @param  signature           If loan payer is an account, then this must be the tightly-packed
+     *                             ECDSA V/R/S parameters from signing the loan hash. If loan payer
+     *                             is a smart contract, these are arbitrary bytes that the contract
+     *                             will recieve when choosing whether to approve the loan.
+     * @param  order               Order object to be passed to the exchange wrapper
+     * @return                     Amount of owedTokens pulled from the lender
+     */
+    function mintShortTokens(
+        bytes32    positionId,
+        address[7] addresses,
+        uint256[8] values256,
+        uint32[2]  values32,
+        bytes      signature,
+        bytes      order
+    )
+        external
```

Comment:
Make `nonReentrant`

For history: we considered an attack where the initial depositor sends ETH which is then wrapped to WETH, and then somewhere a malicious contract is called which calls back into this and effectively steals the WETH from the initial caller by increasing position by 0 and being sent all the `leftoverEth` on line 146